### PR TITLE
Idea for cluster upgrades.

### PR DESCRIPTION
**Pull request by erictune**
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/27080)

### CODE REVIEW

File: docs/proposals/cluster-upgrades.md [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+Kubernetes project wants to release often (~4 releases per year currently, and many people would like to get to 6 or 12).  In each release, these things might change:
+  - version of kubelet required by control plane
+  - version of docker required by kubelet
+  - apiserver, controller-manager, kubelet flags
+  - storage representation of API objects
+  - kernel version and base OS config requirements
+  - existence of added fields on API objects (not captured by apiVersion) depended on by controllers, addons, etc.
+
+It is practical for the Kubernetes project to test forwards and backwards compatibility for
```
Comment (nikhiljindal) [inline docs/proposals/cluster-upgrades.md]: incomplete sentence

File: docs/proposals/cluster-upgrades.md [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+Stable API types (v1, v2, etc) will to supported for fairly long times (e.g. 1 year, and probably longer).  This doc is not about that.
+
+Critical security patches will be backported to all recetnt minor releases (maybe 1 year?  TBD).  This doc is not about that.
+
+It is about "full upgrades" of the programs that make up the cluster (docker, kubelet, apiserver, controller-manager, addons, and their respective versions and flags and many options and input files).
+in order to get new features and performance improvements.
+
+User attitudes to upgrades differ.  Here are three user archetypes:
+  - "yearly upgrader": In my company, we only upgrade software once per year, because that is how we have always done it.  Unless there is a critical security fix.
```
Comment (aronchick) [inline docs/proposals/cluster-upgrades.md]: There's other reasons to be slow to upgrade (things are working fine now; we have deep acceptance process for new software; impacts lots of groups; etc)

File: docs/proposals/cluster-upgrades.md [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+  - "eager upgrader": I want to upgrade my cluster to every new kubernetes release.  I enjoy reading the release notes for each new release of kubernetes with a cup of tea and a chocolate croissant.  Then I try out any interesting new apiserver flags before lunch.
```
Comment (aronchick) [inline docs/proposals/cluster-upgrades.md]: s/chocolate croissant/pain au chocolat/

File: docs/proposals/cluster-upgrades.md [hunk 4]
```diff
@@ -0,0 +1,104 @@
+# Cluster Upgrades
+
+This is a position statement on Cluster Upgrades.  Submitted for feedback.
```
Comment (aronchick) [inline docs/proposals/cluster-upgrades.md]: When I read "cluster" i thought "master + nodes" - is that what this is?

File: docs/proposals/cluster-upgrades.md [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+It is about "full upgrades" of the programs that make up the cluster (docker, kubelet, apiserver, controller-manager, addons, and their respective versions and flags and many options and input files).
+in order to get new features and performance improvements.
+
+User attitudes to upgrades differ.  Here are three user archetypes:
+  - "yearly upgrader": In my company, we only upgrade software once per year, because that is how we have always done it.  Unless there is a critical security fix.
+  - "eager upgrader": I want to upgrade my cluster to every new kubernetes release.  I enjoy reading the release notes for each new release of kubernetes with a cup of tea and a chocolate croissant.  Then I try out any interesting new apiserver flags before lunch.
+  - "hosted upgrader": I use a hosted Kubernetes (e.g. GKE) and trust my hosting provider to keep my cluster up to date.  I just want it to keep working without any effort.  I do not know or care what flags my apiserver has.
+We need to have a story for all these types of users.
+
```
Comment (aronchick) [inline docs/proposals/cluster-upgrades.md]: See above in re: "master + nodes" question

File: docs/proposals/cluster-upgrades.md [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+ - e2e tests that ensure upgrade is possible.
+
+If Kubernetes only provides "single minor version" upgrades, then:
+  - the "yearly upgrader" customer is unhappy, because she needs to do 4 or more consecutive upgrade steps (since a jump from say 1.5 to 1.9 is unlikely to work).
+  - the "eager upgrader" and "hosted upgrader" are happy with this.
+  - the kubernetes team is happy because the upgrade testing problem is tractible
+
+### Multiple-minor-version supported
+
+If kubernetes team has to support multiple minor version upgrades, then the testing matrix gets very large.  Ick.
```
Comment (aronchick) [inline docs/proposals/cluster-upgrades.md]: Could we provide the ability to jump under the hood (e.g. 1.5 -> 1.9 is actually 1.5 -> 1.6 -> 1.7 -> 1.8 -> 1.9)?

File: docs/proposals/cluster-upgrades.md [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+  - "hosted upgrader": I use a hosted Kubernetes (e.g. GKE) and trust my hosting provider to keep my cluster up to date.  I just want it to keep working without any effort.  I do not know or care what flags my apiserver has.
+We need to have a story for all these types of users.
+
+Kubernetes project wants to release often (~4 releases per year currently, and many people would like to get to 6 or 12).  In each release, these things might change:
+  - version of kubelet required by control plane
+  - version of docker required by kubelet
+  - apiserver, controller-manager, kubelet flags
+  - storage representation of API objects
+  - kernel version and base OS config requirements
+  - existence of added fields on API objects (not captured by apiVersion) depended on by controllers, addons, etc.
```
Comment (vishh) [inline docs/proposals/cluster-upgrades.md]: There is also the plugins/extension API for storage, networking, monitoring, logging, etc.

**Pull request closed without merging**